### ISSUE TITLE
MVP implementation of a `masked_scan_combinator`

### ISSUE THREAD

**Issue by georgematheos** [genjax-community/genjax#1184]
```python
import jax
import jax.numpy as jnp
import genjax
from genjax import gen, Mask
genjax.pretty()

def masked_scan_combinator(step, **scan_kwargs):
    """
    Given a generative function `step` so that `step.scan(n=N)` is valid,
    return a generative function accepting an input
    `(initial_state, masked_input_values_array)` and returning a pair
    `(masked_final_state, masked_returnvalue_sequence)`.
    This operates similarly to `step.scan`, but the input values can be masked.
    """
    mstep = step.mask().dimap(
        pre=lambda masked_state, masked_inval: (
            jnp.logical_and(masked_state.flag, masked_inval.flag),
            masked_state.value,
            masked_inval.value
        ),
        post=lambda args, masked_retval: (
            Mask(masked_retval.flag, masked_retval.value[0]),
            Mask(masked_retval.flag, masked_retval.value[1])
        )
    )

    # This should be given a pair (
    #     Mask(True, initial_state), 
    #     Mask(bools_indicating_active, input_vals)
    # ).
    # It will output a pair (masked_final_state, masked_returnvalue_sequence).
    scanned = mstep.scan(**scan_kwargs)

    scanned_nice = scanned.dimap(
        pre=lambda initial_state, masked_input_values: (
            Mask(True, initial_state),
            Mask(masked_input_values.flag, masked_input_values.value)
        ),
        post=lambda args, retval: retval
    )

    return scanned_nice

### Usage example ###
@gen
def step(state, sigma):
    new_x = genjax.normal(state, sigma) @ "x"
    return (new_x, new_x + 1)

trace = masked_scan_combinator(step, n=10).simulate(
    jax.random.PRNGKey(20), (
        2.,
        Mask(jnp.arange(10) < 5, 2. + jnp.arange(10, dtype=float))
    )
)
```

**Comment by georgematheos** [comment]
For more specialized cases, there could be even nicer variants of this.

For example, I'd love a `step_gen_fn.unfold(max_length=...)(initial_state, actual_length)` for `step_gen_fn` accepting a single `state` argument and outputting a pair `(next_state, retval_for_this_step)`.  The unfold combinator could return `masked_final_state, masked_retval_list`.  The first `actual_length` elements of `masked_retval_list` would have flag=True; the rest would have flag=False.  The flag for `masked_final_state` would be `True` iff `max_length == actual_length`.

But I haven't thought too much about the most ergonomic interfaces, so I would not be surprised if there are even nicer designs.

**Comment by georgematheos** [comment]
Update: edited the code snippet so it no longer tries to instantiate a 0-variance gaussian. The error persists.

Also -  note from Rif, which I agree with - it would probably be better if the `masked_scan_combinator` exposed a gen fn accepting a single `length` argument, rather than letting users mask on a per-timestep basis.  We do not have in mind any use cases where it is useful to mask on a per-timestep basis.

**Comment by littleredcomputer** [comment]
The following adjusted code now works in main. What has changed: We add some Flag() wrappers where needed, and use the Flag logical operation.

```python
def MaskedScanCombinator(step, **scan_kwargs):
    """
    Given a generative function `step` so that `step.scan(n=N)` is valid,
    return a generative function accepting an input
    `(initial_state, masked_input_values_array)` and returning a pair
    `(masked_final_state, masked_returnvalue_sequence)`.
    This operates similarly to `step.scan`, but the input values can be masked.
    """
    mstep = step.mask().dimap(
        pre=lambda masked_state, masked_inval: (
            masked_state.flag.and_(masked_inval.flag),
            masked_state.value,
            masked_inval.value
        ),
        post=lambda args, masked_retval: (
            genjax.Mask(masked_retval.flag, masked_retval.value[0]),
            genjax.Mask(masked_retval.flag, masked_retval.value[1])
        )
    )

    # This should be given a pair (
    #     Mask(True, initial_state),
    #     Mask(bools_indicating_active, input_vals)
    # ).
    # It will output a pair (masked_final_state, masked_returnvalue_sequence).
    scanned = mstep.scan(**scan_kwargs)

    scanned_nice = scanned.dimap(
        pre=lambda initial_state, masked_input_values: (
            genjax.Mask(Flag(True), initial_state),
            genjax.Mask(masked_input_values.flag, masked_input_values.value)
        ),
        post=lambda args, retval: retval
    )

    return scanned_nice

### Usage example ###
@gen
def step(state, sigma):
    new_x = genjax.normal(state, sigma) @ "x"
    return (new_x, new_x + 1)

trace = MaskedScanCombinator(step, n=10).simulate(
    jax.random.PRNGKey(20), (
        2.,
        genjax.Mask(Flag(jnp.arange(10) < 5), 2. + jnp.arange(10, dtype=float))
    )
)
```

**Comment by sritchie** [comment]
If we go this route @littleredcomputer we should make it easier to create flags. It's a funky import now. Or we could have `mask` accept a `BoolArray` also and build its own flag in `__init__`?

**Comment by sritchie** [comment]
Just adding some notes here:

We've also got @derifatives ' and Yoni's `masked_iterate_final`

```python
def masked_iterate_final(step: GenerativeFunction[Carry]) -> GenerativeFunction[Carry]:
    def pre_fn(state: Carry, flag: Flag) -> tuple[Flag, Carry]:
        return flag, state

    def post_fn(_, ret: Mask[Carry]) -> Carry:
        return ret.value
        
    return step.mask().dimap(pre=pre_fn, post=post_fn).reduce()
```

which reduces across an initial (unmasked) value and a `BoolArray` and internally masks each element seen in the reduce.

**Comment by sritchie** [comment]
Here is a typed version of @georgematheos 's code:

```python
X = TypeVar("X")

def masked_scan(
    step: GenerativeFunction[tuple[Carry, Y]], *, n: Int | None = None
) -> GenerativeFunction[tuple[Mask[Carry], Mask[Y]]]:
    """
    Given a step of type
    
    (carry, x) -> (carry, y),
    
    returns a new function of type

    (mask[carry], [mask[x]]) -> (mask[carry], [mask[y]])
    """
    def flag_and(a: Flag | Diff[Flag], b: Flag | Diff[Flag]) -> Flag:
        assert not isinstance(a, Diff) and not isinstance(b, Diff)
        return FlagOp.and_(a, b)

    def pre(
        masked_carry: Mask[Carry], masked_x: Mask[X]
    ) -> tuple[Flag, Carry, X]:
        return (
            flag_and(masked_carry.flag, masked_x.flag),
            masked_carry.value,
            masked_x.value
        )

    def post(
        _, masked_retval: Mask[tuple[Carry, Y]]
    ) -> tuple[Mask[Carry], Mask[Y]]:
        flag = masked_retval.flag
        carry, y = masked_retval.value
        return (Mask(flag, carry), Mask(flag, y))

    return step.mask().dimap(
        pre=pre, post=post
    ).scan(n=n)
```

**Comment by sritchie** [comment]
Here is another suggestion:

```python
def masked_scan(
    step: GenerativeFunction[tuple[Carry, Y]]
) -> GenerativeFunction[tuple[Carry, Y]]:
    """
    Given a step of type
    
    (carry, x) -> (carry, y),
    
    returns a new function of type

    (carry, [mask[x]]) -> (carry, [y]])
    """
    def pre(carry: Carry, mask: Mask[X]) -> tuple[Flag, Carry, X]:
        return mask.primal_flag(), carry, mask.value

    def post(_, masked_retval: Mask[tuple[Carry, Y]]) -> tuple[Carry, Y]:
        return masked_retval.value

    return step.mask().dimap(
        pre=pre, post=post
    ).scan()

@genjax.gen
def step(state, sigma):
    new_x = genjax.normal(state, sigma) @ "x"
    return (new_x, new_x + 1)

key = jax.random.PRNGKey(20)    

trace = masked_scan(step).simulate(
    key, (
        2.,
        Mask(jnp.arange(10) < 5, 2. + jnp.arange(10, dtype=float))
    )
)
trace.get_choices()
```

**Comment by sritchie** [comment]
The question here is, do we want to have the mask pop out of the returned values or not. The first version actually propagates the carry's flag, and if it goes false, the whole thing goes false. Is that what we want?

Or do we ONLY want the flags on the input to handle the masking on the output?

**Comment by sritchie** [comment]
@georgematheos , I am going to close this issue, in favor of the discussion that we started at the Con. @femtomc is going to write up notes from his discussions with you, @alex-lew, @arijitnoobstar and others at https://linear.app/chi-fro/issue/GEN-588/write-up-the-user-led-ideas-behind-requests-for-masking; then we'll convene a meeting of the minds where we can all stare at this together and come up with implementations that everyone feels good about.

We are not going to go do this in a vacuum! I am just becoming less convinced that an approach that feeds masks through the modeling layer is the user-friendly combinator we want to end up with... let's figure it out together!
